### PR TITLE
*: add watch for services' primary address

### LINKED ISSUE #6143: Refine the implementation of (*Server).GetServicePrimaryAddr
## Development Task

https://github.com/hnes/pd/blob/a29ff0031551520753af6500acb44e7c6fecf1b1/server/server.go#L1668-L1680

We should use cache mechanism to prevent the etcd from being overwhelmed under some unexpected circumstances.

### PR DESCRIPTION
**Pull request by lhy1024**
### What problem does this PR solve?

Issue Number: Close #6143  Ref https://github.com/tikv/pd/issues/5836.

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Unit test

### Release note

```release-note
None.
```

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -1665,17 +1673,56 @@ func (s *Server) UnmarkSnapshotRecovering(ctx context.Context) error {
 	return err
 }
 
+type servicePrimary struct {
```
Comment (rleungx) [inline server/server.go:1676]: Why do we need this struct?

Comment (lhy1024) [reply server/server.go:1676]: only keep result is unchanged, how about removing error and output it in log if it is not nil?

File: server/server.go [hunk 2]
```diff
@@ -529,6 +531,12 @@ func (s *Server) startServerLoop(ctx context.Context) {
 	go s.serverMetricsLoop()
 	go s.tsoAllocatorLoop()
 	go s.encryptionKeyManagerLoop()
+
+	if s.IsAPIServiceMode() {
+		s.serverLoopWg.Add(2)
+		go s.watchServicePrimaryKeyLoop("tso")
+		go s.watchServicePrimaryKeyLoop("resource-manager")
```
[the remark this answers is not in the document]
Comment (lhy1024) [reply server/server.go:538]: I am not sure whether we have other service need to watch,scheduler?

File: tests/mcs/tso/server_test.go [hunk 3]
```diff
@@ -120,6 +120,12 @@ func (suite *tsoServerTestSuite) TestTSOServerStartAndStopNormally() {
 }
 
 func (suite *tsoServerTestSuite) TestTSOServerRegister() {
+	for i := 0; i < 3; i++ {
```
Comment (rleungx) [inline tests/mcs/tso/server_test.go:123]: Why add a loop?

**Review by rleungx — COMMENTED** [review]
Please add a test for this change.

File: tests/mcs/tso/server_test.go (hunk 3 above)
Comment (lhy1024) [reply tests/mcs/tso/server_test.go:123]: check whether can get a new address when the primary update

Comment (rleungx) [reply tests/mcs/tso/server_test.go]: Why adding a loop can do the check?

Comment (lhy1024) [reply tests/mcs/tso/server_test.go]: `checkServerRegister` will add a tso server or resourse_manager server, so we can check whether the primary changed.

**Review by rleungx — APPROVED** [review]

**Comment by lhy1024** [comment]
cc @rleungx add retry and ptal

File: server/server.go [hunk 4]
```diff
@@ -1674,9 +1679,16 @@ func (s *Server) UnmarkSnapshotRecovering(ctx context.Context) error {
 
 // GetServicePrimaryAddr returns the primary address for a given service.
 func (s *Server) GetServicePrimaryAddr(serviceName string) (string, bool) {
-	// TODO: add ping to check if the primary is alive and retry.
-	if v, ok := s.servicePrimaryMap.Load(serviceName); ok {
-		return v.(string), true
+	// TODO: add ping to check if the primary is alive.
+	for i := 0; i < maxRetryTimes; i++ {
```
Comment (rleungx) [inline server/server.go:1683]: Does it work? Have you tried it?

File: server/server.go [hunk 5]
```diff
@@ -98,6 +99,11 @@ const (
 	PDMode = "PD"
 	// APIServiceMode represents that server is in API service mode.
 	APIServiceMode = "API service"
+
+	// maxRetryTimes is the max retry times for getting primary addr.
+	maxRetryTimes = 50
+	// retryInterval is the interval to retry.
+	retryInterval = 100 * time.Millisecond
```
Comment (JmPotato) [inline server/server.go:106]: Will 5 seconds kind of long? What about 3 seconds?

File: server/server.go (hunk 4 above)
Comment (lhy1024) [reply server/server.go:1683]: yes, it can run in hack tiup cluster.

File: server/server.go [hunk 6]
```diff
@@ -1666,16 +1678,58 @@ func (s *Server) UnmarkSnapshotRecovering(ctx context.Context) error {
...
+
+	watchChan := s.client.Watch(ctx, serviceKey, clientv3.WithPrefix())
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info("server is closed, exist watch service primary addr loop", zap.String("service", serviceName))
+			return
+		case res := <-watchChan:
+			for _, event := range res.Events {
+				if event.Type != clientv3.EventTypePut {
```
Comment (rleungx) [inline server/server.go:1722]: Do we need to handle the delete event?

File: tests/mcs/discovery/register_test.go [hunk 7]
```diff
@@ -66,38 +68,87 @@ func (suite *serverRegisterTestSuite) TearDownSuite() {
...
 
 func (suite *serverRegisterTestSuite) TestServerRegister() {
-	suite.checkServerRegister("tso")
-	suite.checkServerRegister("resource_manager")
+	// test register, primary and unregister when start tso and resource-manager with only one server
+	for i := 0; i < 3; i++ {
+		suite.checkServerRegister("tso")
+	}
+	for i := 0; i < 3; i++ {
+		suite.checkServerRegister("resource_manager")
```
[the remark this answers is not in the document]
Comment (lhy1024) [reply tests/mcs/discovery/register_test.go:76]: I will do it in another pr after #6155 and #6130 merged

File: server/server.go [hunk 8]
```diff
@@ -1666,16 +1673,51 @@ func (s *Server) UnmarkSnapshotRecovering(ctx context.Context) error {
...
-func (s *Server) GetServicePrimaryAddr(ctx context.Context, serviceName string) (bool, string, error) {
+func (s *Server) GetServicePrimaryAddr(ctx context.Context, serviceName string) (string, bool) {
+	// TODO: add ping to check if the primary is alive and retry.
+	if v, ok := s.servicePrimaryMap.Load(serviceName); ok {
+		return v.(string), true
+	}
+	return "", false
+}
+
+func (s *Server) watchServicePrimaryAddrLoop(serviceName string) {
```
[the remark this answers is not in the document]
Comment (lhy1024) [reply server/server.go:1698]: we will only watched this key with `s.clusterID` after #6166

**Review by binshi-bing — APPROVED** [review]
left few comments

**Review by JmPotato — APPROVED** [review]

**Pull request merged**